pkg/controller/generic/cleanup/cleanup.go

```diff
@@ -0,0 +1,348 @@
+			err := ctrl.processInput(ctx, r, logger, iter.Value())
+			if err != nil {
+				multiErr = errors.Join(multiErr, err)
+			}
+		}
+
+		if multiErr != nil {
+			return fmt.Errorf("cleanup controller %q failed: %w", ctrl.Name(), multiErr)
+		}
+	}
+}
+
+func (ctrl *Controller[I]) processInput(ctx context.Context, r controller.Runtime, logger *zap.Logger, inputElem I) error {
+	l := logger.With(
+		zap.String("id", inputElem.Metadata().ID()),
+		zap.String("resource", inputElem.ResourceDefinition().Type),
+	)
+
+	switch inputElem.Metadata().Phase() {
+	case resource.PhaseTearingDown:
+		if !inputElem.Metadata().Finalizers().Has(ctrl.Name()) {
+			return nil
+		}
+
+		err := ctrl.handler.FinalizerRemoval(ctx, r, logger, inputElem)
+
+		switch {
+		case xerrors.TagIs[SkipReconcileTag](err):
+			// skip this resource
+			return nil
+
+		case err != nil:
+			return fmt.Errorf(
+				"error handling finalizer removal from resource %q with id %q: %w",
+				inputElem.ResourceDefinition().Type,
+				inputElem.Metadata().ID(),
+				err,
+			)
+		}
+
+		if err := r.RemoveFinalizer(ctx, inputElem.Metadata(), ctrl.Name()); err != nil {
+			return fmt.Errorf(
+				"error removing finalizer from resource %q with id %q: %w",
+				inputElem.ResourceDefinition().Type,
+				inputElem.Metadata().ID(),
+				err,
+			)
+		}
+
+		l.Info("removed finalizer")
+
+	case resource.PhaseRunning:
+		if inputElem.Metadata().Finalizers().Has(ctrl.Name()) {
+			return nil
+		}
+
+		err := r.AddFinalizer(ctx, inputElem.Metadata(), ctrl.Name())
+		if err != nil {
+			return fmt.Errorf(
+				"error adding finalizer on resource %q with id %q: %w",
+				inputElem.ResourceDefinition().Type,
+				inputElem.Metadata().ID(),
+				err,
+			)
+		}
+
+		l.Info("put finalizer on resource")
+	}
+
+	return nil
+}
+
+// handler implements [Handler] interface. Helper type, not intended to be exported.
+type handler[I generic.ResourceWithRD, O generic.ResourceWithRD] struct {
+	finalizerRemoval func(context.Context, controller.Runtime, *zap.Logger, I) error
+	inputs           func() []controller.Input
+	outputs          func() []controller.Output
+}
+
+// FinalizerRemoval implements controller.Handler interface.
+func (h *handler[I, O]) FinalizerRemoval(ctx context.Context, rt controller.Runtime, l *zap.Logger, input I) error {
+	if h.finalizerRemoval != nil {
+		return h.finalizerRemoval(ctx, rt, l, input)
+	}
+
+	return nil
+}
+
+// Inputs implements controller.Controller interface.
+func (h *handler[I, O]) Inputs() []controller.Input {
+	if h.inputs != nil {
+		return h.inputs()
+	}
+
+	return nil
+}
+
+// Outputs implements controller.Controller interface.
+func (h *handler[I, O]) Outputs() []controller.Output {
+	if h.outputs != nil {
+		return h.outputs()
+	}
+
+	return nil
+}
+
+// HasNoOutputs is a helper function to create a [Handler] that waits for all outputs to be removed.
+func HasNoOutputs[O generic.ResourceWithRD, I generic.ResourceWithRD](
```

Comment:
how can we handle multiple outputs here? i.e. what if one resource depends on multiple output resources?

Reply:
Some sort of `combine` should do the trick. I haven't really thought about it.

Reply:
but this seems tricky right now, as input is singular, but outputs are not singular

Reply:
Done in `Combine` func.